android/KMEA/app/src/main/java/com/keyman/engine/KMKeyboard.java

```diff
@@ -184,17 +186,50 @@ protected boolean updateSelectionRange(int selStart, int selEnd) {
         References:
         - https://stackoverflow.com/a/23980211
         - https://android.googlesource.com/platform/frameworks/base/+/152944f/core/java/android/view/inputmethod/InputConnection.java#326
-       */
+
+        We need to ensure that:
+        a) selStart <= selEnd (so "reverse" selection if needed)
+        b) selStart >= 0 and selEnd >= 0
+        c) selStart <= string.length() and selEnd <= string.length()
+        Log if we need to make any adjustments
+      */
+      boolean adjustedValues = false;
+      if (selStart > selEnd) {
+        adjustedSelStart = selEnd;
+        adjustedSelEnd = selStart;
+        adjustedValues = true;
+      }
+      if (adjustedSelStart < 0) {
+        adjustedSelStart = 0;
+        adjustedValues = true;
+      }
+      if (adjustedSelEnd < 0) {
+        adjustedSelEnd = 0;
+        adjustedValues = true;
+      }
+      if (adjustedSelStart > rawText.length()) {
+        adjustedSelStart = rawText.length();
+        adjustedValues = true;
+      }
+      if (adjustedSelEnd > rawText.length()) {
+        adjustedSelEnd = rawText.length();
+        adjustedValues = true;
+      }
+      if (adjustedValues) {
+        // Log when we have to adjust selStart / selEnd
+        KMLog.LogInfo(TAG, KMString.format("Constrained selStart: %d, selEnd: %d to adjustedSelStart: %d, adjustedSelEnd %d",
+          selStart, selEnd, adjustedSelStart, adjustedSelEnd));
+      }
 
       // Count the number of characters which are surrogate pairs.
-      int pairsAtStart = CharSequenceUtil.countSurrogatePairs(rawText.substring(0, selStart), rawText.length());
-      String selectedText = rawText.substring(selStart, selEnd);
+      int pairsAtStart = CharSequenceUtil.countSurrogatePairs(rawText.substring(0, adjustedSelStart), rawText.length());
+      String selectedText = rawText.substring(adjustedSelStart, adjustedSelEnd);
```

Comment:
Dunno if we also need to constrain for selectedText.length()

Reply:
^ If so - I'll prob refactor the constrain code into a helper function. e.g. also unit testable

Reply:
I ended up refactoring it to a helper function. Though selectedString should fit and not need any additional constraining